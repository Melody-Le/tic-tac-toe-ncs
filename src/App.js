import Board from "./components/Board/Board";
import GlobalStyle from "./theme/globalStyle";

function App() {
  return (
    <>
      <GlobalStyle />
      <Board />
    </>
  );
}

export default App;
