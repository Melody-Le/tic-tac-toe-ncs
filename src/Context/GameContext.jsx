import React, { createContext, useContext, useState, useEffect } from "react";
import { GameState, PLAYER_X, PLAYER_O } from "../utils/constant";
import { LOCAL_STORAGE_KEY } from "../utils/constant";
import { defineWinner } from "../utils/fn";

export const GameContext = createContext({});

export function useGame() {
  return useContext(GameContext);
}

export function GameProvider({ children }) {
  const [history, setHistory] = useState(() => {
    const histoyLCS = JSON.parse(
      window.localStorage.getItem(LOCAL_STORAGE_KEY)
    );
    return (
      histoyLCS || [
        {
          step: 0,
          squares: Array(9).fill(null),
          squareIndex: null,
        },
      ]
    );
  });

  const [currentStep, setCurrentStep] = useState(() => {
    const histoyLCS = JSON.parse(
      window.localStorage.getItem(LOCAL_STORAGE_KEY)
    );
    if (histoyLCS) {
      return histoyLCS[histoyLCS?.length - 1].step;
    } else {
      return 0;
    }
  });
  const [gameState, setGameState] = useState(GameState.inProgress);

  const [winningLine, setwinningLine] = useState([]);

  const [player, setPlayer] = useState(PLAYER_X);

  const currentSquares = history[currentStep]?.squares;

  const [squareIndexOnHistoryHover, setSquareIndexOnHistoryHover] =
    useState(null);

  const handlePlay = (nextSquares, clickedSquareIndex) => {
    const nextHistory = [
      ...history.slice(0, currentStep + 1),
      {
        step: currentStep + 1,
        squares: nextSquares,
        squareIndex: clickedSquareIndex,
      },
    ];
    setHistory(nextHistory);
    setCurrentStep(nextHistory.length - 1);
    player === PLAYER_X ? setPlayer(PLAYER_O) : setPlayer(PLAYER_X);
    window.localStorage.setItem(LOCAL_STORAGE_KEY, nextHistory);
  };
  const handleRestart = () => {
    localStorage.clear();
    setHistory([
      {
        step: 0,
        squares: Array(9).fill(null),
        squareIndex: null,
      },
    ]);
    setGameState(GameState.inProgress);
    setPlayer(PLAYER_X);
    setwinningLine([]);
    setCurrentStep(0);
    setSquareIndexOnHistoryHover(null);
  };
  const handleHistoryMouseOver = (index) => {
    setSquareIndexOnHistoryHover(index);
  };
  const handleHistoryMouseOut = () => {
    setSquareIndexOnHistoryHover(null);
  };

  const handleHistoryStepClick = (index) => {
    jumpTo(index - 1);
    setSquareIndexOnHistoryHover(null);
  };
  const jumpTo = (nextStep) => {
    setCurrentStep(nextStep);
    setPlayer(nextStep % 2 === 0 ? PLAYER_X : PLAYER_O);
    const nextHistory = [...history.slice(0, nextStep + 1)];
    setHistory(nextHistory);
    window.localStorage.setItem(LOCAL_STORAGE_KEY, nextHistory);
  };

  const handleBackAStep = () => {
    currentStep > 1 && jumpTo(currentStep - 2);
  };

  const handleSquareClick = (index) => {
    const nextSquare = [...currentSquares];
    if (nextSquare[index] || gameState !== GameState.inProgress) return;
    nextSquare[index] = player === PLAYER_X ? "X" : "O";
    handlePlay(nextSquare, index);
  };

  useEffect(() => {
    window.localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(history));
  }, [history]);

  useEffect(() => {
    const winner = defineWinner(currentSquares)?.winner;
    switch (winner) {
      case "X":
        setGameState(GameState.playerOWins);
        setwinningLine(defineWinner(currentSquares)?.winnningPositons);
        break;
      case "O":
        setwinningLine(defineWinner(currentSquares)?.winnningPositons);
        setGameState(GameState.playerOWins);
        break;
      case "draw":
        setGameState(GameState.draw);
        break;
      default:
        setGameState(GameState.inProgress);
    }
  }, [currentSquares]);
  const value = {
    history,
    gameState,
    currentSquares,
    currentStep,
    player,
    winningLine,
    squareIndexOnHistoryHover,
    jumpTo,
    handleRestart,
    handleBackAStep,
    handleHistoryMouseOver,
    handleHistoryMouseOut,
    handleHistoryStepClick,
    handleSquareClick,
  };
  return <GameContext.Provider value={value}>{children}</GameContext.Provider>;
}
