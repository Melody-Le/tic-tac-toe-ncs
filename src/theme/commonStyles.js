export const BREAKPONITS = {
  small: `@media (min-width: 576px)`,
  tablet: `@media (min-width: 768px)`,
  large: `@media (min-width: 992px)`,
};
export const squareWrapperSize = 20;
export const gap = { sm: 0.25, md: 0.5, lg: 1, xl: 2 };
export const padding = { sm: 0.5, md: 1, lg: 2, xl: 3 };
export const borderRadius = 0.5;
export const fontSize = { sm: 1, md: 2, lg: 3, xl: 4 };
