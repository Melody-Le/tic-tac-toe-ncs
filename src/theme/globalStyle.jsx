import { createGlobalStyle } from "styled-components";

const GlobalStyle = createGlobalStyle`
  html {
    font-size: 100%;
    box-sizing: border-box;
  }
  body{
    font-family:-apple-system, BlinkMacSystemFont, Roboto, sans-serif;
    margin: 0;
    padding: 0;
    display: grid;
    place-content: center;
    min-height: 100vh;
  }
  * {
    margin: 0;
    padding: 0;
    box-sizing: inherit;
    margin: 0 auto;
  }
`;
export default GlobalStyle;
