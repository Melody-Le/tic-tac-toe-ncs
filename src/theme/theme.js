export const theme = {
  darkBlue: "#0f1b21",
  primary: "#26ffcb",
  secondPrimary: "hsl(0, 100%, 74%)",
  secondDarkBlue: "#1f3540",
  secondHightlight: "#436f85",
  highlight: "#fcf261",
  bright: "#ffffff",
  secondBright: "#26ffcb",
};
