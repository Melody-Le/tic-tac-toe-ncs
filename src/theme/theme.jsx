const Theme = {
  colors: {
    darkBlue: "#2C465C",
    teal: "#7ABEBB",
    coral: "#FF7A7A",
    white: "#ffffff",
  },
};
