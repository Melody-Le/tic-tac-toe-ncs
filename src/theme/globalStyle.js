import { createGlobalStyle } from "styled-components";
import { theme } from "./theme";
import { padding, BREAKPONITS } from "./commonStyles";

const GlobalStyle = createGlobalStyle`
  html {
    font-size: 100%;
    box-sizing: border-box;
    color: ${theme.secondBright};

  }
  body{
    font-family:-apple-system, BlinkMacSystemFont, Roboto, sans-serif;
    margin: 0;
    padding: ${padding.md}rem; 
    display: flex;
    justify-content: center; 
    min-height: 100vh;
    background-color:${theme.darkBlue};

    ${BREAKPONITS.tablet} {
      padding: ${padding.lg}rem; 
    }
  }
  * {
    margin: 0;
    padding: 0;
    box-sizing: inherit;
  }
  button {
    all: unset;
    cursor: pointer;
    &:hover {
      background-color: ${theme.highlight};
      color: ${theme.secondDarkBlue};
    }
  }

`;
export default GlobalStyle;
