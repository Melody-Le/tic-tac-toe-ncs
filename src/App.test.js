import React from "react";
import App from "./App";
import { theme } from "./theme/theme";
import { GameProvider } from "./Context/GameContext";
import { render } from "@testing-library/react";

it("matches snapshot", () => {
  const container = render(
    <GameProvider theme={theme}>
      <App />
    </GameProvider>
  );
  expect(container).toMatchSnapshot();
  // expect(screen.getByTestId("board")).toBeInTheDocument();
});
