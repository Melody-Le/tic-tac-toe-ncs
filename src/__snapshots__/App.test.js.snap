// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdfCDU ksWkj"
      >
        <div
          class="sc-gsTDqH iYxtmk"
        >
          <button
            class="sc-hKgKIp sc-eCstZk leDbbK ijgNJl"
          >
            Restart
          </button>
          <div
            class="sc-dlfmHC eCzLGW"
            data-testid="playerTurnAndStep"
          >
            STEP 
            1
             : 
            X
             TURN
          </div>
        </div>
        <div
          class="sc-iBPTVF gIXXFF"
          data-testid="board"
        >
          <button
            class="sc-gKsecS cuIVwp"
            data-testid="cell-0"
            role="gridcell"
          />
          <button
            class="sc-gKsecS cuIVwp"
            data-testid="cell-1"
            role="gridcell"
          />
          <button
            class="sc-gKsecS cuIVwp"
            data-testid="cell-2"
            role="gridcell"
          />
          <button
            class="sc-gKsecS cuIVwp"
            data-testid="cell-3"
            role="gridcell"
          />
          <button
            class="sc-gKsecS cuIVwp"
            data-testid="cell-4"
            role="gridcell"
          />
          <button
            class="sc-gKsecS cuIVwp"
            data-testid="cell-5"
            role="gridcell"
          />
          <button
            class="sc-gKsecS cuIVwp"
            data-testid="cell-6"
            role="gridcell"
          />
          <button
            class="sc-gKsecS cuIVwp"
            data-testid="cell-7"
            role="gridcell"
          />
          <button
            class="sc-gKsecS cuIVwp"
            data-testid="cell-8"
            role="gridcell"
          />
        </div>
        <div
          class="sc-fubDmA enJhvo"
        >
          <button
            class="sc-pGaPU fulHvW"
            data-testid="backAStepBtn"
          >
            <svg
              fill="currentColor"
              height="2rem"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="2rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.82843 6.99955L8.36396 9.53509L6.94975 10.9493L2 5.99955L6.94975 1.0498L8.36396 2.46402L5.82843 4.99955H13C17.4183 4.99955 21 8.58127 21 12.9996C21 17.4178 17.4183 20.9996 13 20.9996H4V18.9996H13C16.3137 18.9996 19 16.3133 19 12.9996C19 9.68584 16.3137 6.99955 13 6.99955H5.82843Z"
              />
            </svg>
          </button>
          <button
            class="sc-jrAGKZ cidbKW"
            data-testid="showHistoryButton"
          >
            SHOW HISTORY
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdfCDU ksWkj"
    >
      <div
        class="sc-gsTDqH iYxtmk"
      >
        <button
          class="sc-hKgKIp sc-eCstZk leDbbK ijgNJl"
        >
          Restart
        </button>
        <div
          class="sc-dlfmHC eCzLGW"
          data-testid="playerTurnAndStep"
        >
          STEP 
          1
           : 
          X
           TURN
        </div>
      </div>
      <div
        class="sc-iBPTVF gIXXFF"
        data-testid="board"
      >
        <button
          class="sc-gKsecS cuIVwp"
          data-testid="cell-0"
          role="gridcell"
        />
        <button
          class="sc-gKsecS cuIVwp"
          data-testid="cell-1"
          role="gridcell"
        />
        <button
          class="sc-gKsecS cuIVwp"
          data-testid="cell-2"
          role="gridcell"
        />
        <button
          class="sc-gKsecS cuIVwp"
          data-testid="cell-3"
          role="gridcell"
        />
        <button
          class="sc-gKsecS cuIVwp"
          data-testid="cell-4"
          role="gridcell"
        />
        <button
          class="sc-gKsecS cuIVwp"
          data-testid="cell-5"
          role="gridcell"
        />
        <button
          class="sc-gKsecS cuIVwp"
          data-testid="cell-6"
          role="gridcell"
        />
        <button
          class="sc-gKsecS cuIVwp"
          data-testid="cell-7"
          role="gridcell"
        />
        <button
          class="sc-gKsecS cuIVwp"
          data-testid="cell-8"
          role="gridcell"
        />
      </div>
      <div
        class="sc-fubDmA enJhvo"
      >
        <button
          class="sc-pGaPU fulHvW"
          data-testid="backAStepBtn"
        >
          <svg
            fill="currentColor"
            height="2rem"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="2rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.82843 6.99955L8.36396 9.53509L6.94975 10.9493L2 5.99955L6.94975 1.0498L8.36396 2.46402L5.82843 4.99955H13C17.4183 4.99955 21 8.58127 21 12.9996C21 17.4178 17.4183 20.9996 13 20.9996H4V18.9996H13C16.3137 18.9996 19 16.3133 19 12.9996C19 9.68584 16.3137 6.99955 13 6.99955H5.82843Z"
            />
          </svg>
        </button>
        <button
          class="sc-jrAGKZ cidbKW"
          data-testid="showHistoryButton"
        >
          SHOW HISTORY
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
